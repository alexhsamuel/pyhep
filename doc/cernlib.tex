\section{\module{hep.cernlib} -- CERNLIB interface}

\declaremodule{extension}{hep.cernlib}
\modulesynopsis{CERNLIB interface}

The \module{hep.cernlib} module provides access from Python to features
in CERNLIB, the CERN program library for high energy physics.  The
applicable components of CERNLIB are linked into \pyhep and need not be
provided separately.

\subsection{\module{hep.cernlib.hbook} -- HBOOK}

\declaremodule{extension}{hep.cernlib.hbook}
\modulesynopsis{HBOOK interface}

HBOOK is a system for managing histograms and ntuples.  It implements a
file format for storing these in structured disk files.  \pyhep provides
read and write access to histograms and ntuples stored in HBOOK files.

The module \module{hep.cernlib.hbook} provides these functions:

\begin{funcdesc}{open}{path\optional{, mode="r", record_length=1024,
purge_cycles=1}}
 Create or open an HBOOK file at \var{path}.  

 \var{mode} is the access mode, analogous to Python's built-in
 \function{file} and \function{open} functions.  If \code{"r"}, opens an
 existing HBOOK file for reading.  If \code{"r+"}, \code{"w"},
 \code{"a"}, or \code{"a+"}, opens an existing file for reading and
 writing, or creates the file if it doesn't exist.  If \code{"w+"},
 creates a file or overwrites an existing file.

 \var{record_length} is the RZ record length to use for this file.  

 If \var{purge_cycles} is set to a true value and the file is open for
 writing, the entire file will be purged when it is closed.
\end{funcdesc}

The return value from \function{open} is a \class{File} object.  The
file remains open until it is deleted.  

\subsubsection{Paths, IDs, and cycles}

In \pyhep, all paths in HBOOK files are specified relative to the root
directory of the HBOOK file.  The root directory itself is represented
by an empty string, and the path separator is a forward slash.  Note
that there is no notion of working directory or relative paths.

The last component of a path is the title of a histogram, ntuple, or
directory.  HBOOK does not require titles to be unique, even in the same
directory.  However, each entry is also assigned a positive integer ID
number (which is called the RZ ID, because HBOOK's file format is built
on top of the RZ input/output system).  \pyhep's functions for creating
tables and directories and saving histograms will choose an available RZ
ID by default, but you can specify a different value to use (except for
directories).

Thus, HBOOK allows you to place multiple items with the same title in
the same directory, but each must have a unique RZ ID.  When loading or
deleting histograms or ntuples, you may specify either the RZ ID instead
of the title in the path.  For example, a histogram with title MYHIST
and RZ ID 17 in directory DIR/SUBDIR can be accessed as
\code{"DIR/SUBDIR/MYHIST"} or \code{"DIR/SUBDIR/17"}.  When saving
histograms, though, you must specify the title in the path.

Note that because of how HBOOK keeps ntuples in memory while they're
being accessed, you may not access two ntuples with the same RZ ID at
the same time, even if the ntuples are stored in different HBOOK files
or directories.  Be sure to delete the \class{Table} object for one
before creating or loading another.

HBOOK files also support revision cycles; see the HBOOK or RZ
documentation for details.  Unless \function{hbook.open} is invoked with
\code{purge_cycles=0} (or the file is opened read-only), \pyhep
``purges'' the entire file when the file is closed.  This removes cycles
except the most recent for each histogram and ntuple in the file.

\subsubsection{\class{File} objects}

A \class{File} object has the following methods and attributes:

\begin{memberdesc}{writeable}
 True if the file is open for writing as well as reading.
\end{memberdesc}

\begin{methoddesc}{listdir}{\optional{path}}
 Return the contents of the directory specified by \var{path}.  The
 return value is a sequence of \code{(id, type, title)} tuples, where
 \var{id} is the RZ ID of the entry; \var{type} is one of
 \code{"1d~histogram"}, \code{"2d~histogram"}, \code{"ntuple"}, or
 \code{"directory"}; and \var{title} is the title of the entry.
\end{methoddesc}

\begin{methoddesc}{mkdir}{path}
 Create a new directory at \var{path}.
\end{methoddesc}

\begin{methoddesc}{rmdir}{path}
 Remove the directory at \var{path}.
\end{methoddesc}

\begin{methoddesc}{save}{path, object\optional{, rz_id}}
 Store a histogram \var{object} at \var{path}.  One- and two-dimensional
 histograms are supported.  An unused RZ ID is chosen automatically,
 unless \var{rz_id} is specified or \var{object} has an \member{rz_id}
 attribute.  The function returns the RZ ID under which 'object' was
 saved. 
\end{methoddesc}

\begin{methoddesc}{load}{path}
 Load a histogram or connect to an ntuple at \var{path}.  If \var{path}
 is a one- or two-dimensional histogram, returns a histogram object with
 its contents.  If \var{path} is an ntuple, returns a table object (see
 \module{hep.table}) accessing it.
\end{methoddesc}

\begin{methoddesc}{createTable}{path, schema\optional{, rz_id, column_wise=1}}
 Create a table as an ntuple at \var{path} in the HBOOK file, using
 columns from \var{schema}, a \class{hep.table.Schema} instance.  If
 \var{rz_id} is not provided, an unused RZ ID is chosen automatically.
 The table is stored as a column-wise ntuple, unless a false value is
 passed for \var{column_wise}.  

 Column-wise ntuples support these column types: \constant{"int32"},
 \constant{"int64"}, \constant{"float32"}, and \constant{"float64"}.

 Row-wise ntuples support only \constant{"float32"} column types.
\end{methoddesc}

\begin{methoddesc}{remove}{path}
 Remove the ntuple or histogram at \var{path}.  Use \method{rmdir} to
 remove a directory.
\end{methoddesc}

\begin{methoddesc}{normpath}{path}
 Return the canonical form of absolute path \var{path} in the HBOOK
 file.
\end{methoddesc}

\begin{methoddesc}{join}{*components}
 Join together \var{components} to form an absolute path.
\end{methoddesc}

\begin{methoddesc}{split}{path}
 Return \code{(dir, name)}, where \var{dir} is the portion of \var{path}
 up to the last directory separator, and \var{name} is the remainder.
\end{methoddesc}

