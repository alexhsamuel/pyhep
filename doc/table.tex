\section{\module{hep.table} --- Tables}

\declaremodule{extension}{hep.table}
\modulesynopsis{Tables.}

A \pyhep \emph{table} is a flat database table with columns of fixed
type and storage size.  Tables are append-only: new rows may be added to
the end of the table, but rows may not be inserted elsewhere, nor can
rows be modified or deleted.  

Each table has a fixed \emph{schema}, which determines the data items
stored in the table and their types.  The schema includes an unordered
set of \emph{columns}, each identified by a name.  Each column has a
fixed data type.  A schema may also include \emph{expressions}, which
are derived quantities which are computed for each row from values of
columns and other expressions for that row.  

The table's data is an ordered sequence of \emph{rows}, each of which
contains exactly one value for each column in the table's schema.

Multiple table implementations are provided.  Implementations provide
different interfaces for creating and opening tables.  However, objects
representing tables and rows satisfy the same protocol across all
implementations.


\subsection{Schema objects}

An instance of \class{hep.table.Schema} represents a table schema.  

\begin{memberdesc}{columns}
 \readonly A sequence of columns in the schema.  Each element is a
 \class{Column} instance.  The order of the columns is unspecified.
\end{memberdesc}

\begin{memberdesc}{expressions}
 \readonly A sequence of expressions in the schema.  Each element is
 expression object (see \module{hep.expr}).  The order of the
 expressions is unspecified.
\end{memberdesc}

\begin{methoddesc}{addColumn}{name, type\optional{, **attributes}}
 Add a column to the sequence.  \var{name} is a character string, and
 \var{type} is a column type (see below).  If additional keyword
 arguments are provided, they are added to the column as attributes.
 New columns should not be added to a schema after the schema has been
 used to create a table.
\end{methoddesc}

\begin{methoddesc}{addExpression}{name, expression\optional{, **attributes}}
 Add an expression to the sequence.  \var{name} is a character string,
 and \var{expression} is an expression as a string or expression object.
 If additional keyword arguments are provided, they are added to the
 expression object as attributes.
\end{methoddesc}

An instance of \class{hep.table.Column} describes a column in a schema.
Do not instantiate this class directly; instead, use the
\method{addColumn} memthod of class \class{Schema}.

\begin{memberdesc}{name}
 \readonly The column's name.
\end{memberdesc}

\begin{memberdesc}{type}
 \readonly The column's data type.
\end{memberdesc}

The column's data type is specified by one of these names:
\begin{tableiii}{ccc}{code}{name}{C type}{Python Type}
  \lineiii{"int8"}{\code{signed char}}{\code{int}}
  \lineiii{"int16"}{\code{signed short}}{\code{int}}
  \lineiii{"int32"}{\code{signed int}}{\code{int}}
  \lineiii{"float32"}{\code{float}}{\code{float}}
  \lineiii{"float64"}{\code{double}}{\code{float}}
\end{tableiii}


An instance of \class{hep.table.Expression} describes a column in a
schema.  Do not instantiate this class directly; instead, use the
\method{addExpression} memthod of class \class{Schema}.

\begin{memberdesc}{name}
 \readonly The expressions's name.
\end{memberdesc}

\begin{memberdesc}{expression}
 \readonly An \class{Expression} instance; see \module{hep.expr}.
\end{memberdesc}

\begin{memberdesc}{formula}
 \readonly A string representation of the expression.
\end{memberdesc}



\subsection{Table objects}

A table object represents an in-memory table or a connection to an
externally-stored table.  In the former case, the table is erased when
the table object is deleted.  In the latter case, the connection is
closed but the table persists when the table object is deleted.

A table object satisfies the Python read-only sequence protocol.  The
sequence elements are the rows of the sequence.  Thus, the
\function{len} function returns the number of rows in the table, and
the \method{__getitem__} method extracts a row by its position in the
sequence. 

In addition, these methods are provided:

\begin{methoddesc}{append}{row\optional{, **kw_args}}
 (Read-write tables only.)  Appends a row to the end of the table.  The
 \var{row} must is a dictionary or other mapping object which maps
 column names to values.  Additional column values may be specified as
 keyword arguments.  If a column is specified both in \var{row} and in
 \var{**kw_args}, the value from latter is used.  A value must be
 specified for each column, and no other names may be specified.
 Returns the row index of the new row.
\end{methoddesc}

\begin{methoddesc}{__iter__}{}
 Returns an iterator over all rows in the table.
\end{methoddesc}

\begin{memberdesc}{rows}
 An interator over all rows in the table.
\end{memberdesc}

\begin{methoddesc}{select}{expr}
 Returns an iterator over rows in the table for which expression
 \var{expr} is true.  \var{expr} may be a string expression formula or
 an expression object.
\end{methoddesc}


\subsection{Row objects}

A row object may represent a row in a table, or a row that is compatible
with a table's schema (but not one of the rows in that table).

The row object satisfies the Python map protocol.  Each key is the name
of a column in the table's schema.  The corresponding map value is the
row's value for that column.  Row objects should be treated as
read-only; it is not possible to modify a table row by changing values
in a row object.

In addition, a row object provides these members and methods:

\begin{methoddesc}{asDict}{}
 Return a new Python dictionary of the column names and corresponding
 values in the row.
\end{methoddesc}

\begin{memberdesc}{index}
 \readonly The index of the row in the table.  The value of this
 attribute is undefined if the row was obtained from the table's
 \method{newRow} method and has not yet been passed to \method{append}.
\end{memberdesc}

\begin{memberdesc}{table}
 \readonly The table to which the row belongs.
\end{memberdesc}


\subsection{Standard implementation}

\pyhep's standard table implementation, in \module{hep.table}, uses a
simple, binary, on-disk representation of the table's data.  Table rows
are automatically loaded from disk as needed.

The \module{hep.table} module provides functions \function{create} and
\function{open} to create and open new tables, respectively.

\begin{funcdesc}{create}{filename, schema}
 Create a new table.  The table is stored in a file named by
 \var{filename}, which is created or overwritten.  The \var{schema}
 argument is a map from column names to \class{Column} instances,
 specifying the columns in the table.

 The return value is a table object, which is opened to the new, empty
 in write mode.
\end{funcdesc}

\begin{funcdesc}{open}{filename\optional{, mode="r"}}
 Open an existing table stored in the file named by \var{filename}.  The
 \var{mode} argument specifies the mode in which to open the table:
 \constant{"r"} to open the table in read-only mode, or \constant{"w"}
 to open the table in read-write mode.

 The return value is a table object.
\end{funcdesc}

