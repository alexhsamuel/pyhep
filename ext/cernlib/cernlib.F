c     ------------------------------------------------------------------
c
c     cernlib.F
c
c     Copyright 2003 by Alex Samuel.  All rights reserved.
c
c     ------------------------------------------------------------------

c     Utilities for CERNLIB support.

c     ------------------------------------------------------------------

      SUBROUTINE hbook_init

      IMPLICIT NONE
      INTEGER PAWC_SIZE
      PARAMETER (PAWC_SIZE = 1000000)
      INTEGER p
      COMMON /PAWC/ p(PAWC_SIZE)

      CALL hlimit(PAWC_SIZE)
      RETURN

      END SUBROUTINE

c     ------------------------------------------------------------------

      SUBROUTINE open_lun(lun, path, status)

      IMPLICIT NONE
      INTEGER lun
      INTEGER status
      CHARACTER*1024 path

      open(unit=lun, file=path, access='SEQUENTIAL', iostat=status)
      RETURN

      END SUBROUTINE

c     ------------------------------------------------------------------

      SUBROUTINE close_lun(lun)

      IMPLICIT NONE
      INTEGER lun

      close(lun)
      RETURN

      END SUBROUTINE

c     ------------------------------------------------------------------

      SUBROUTINE hfnoent(idd, numb)

      IMPLICIT NONE
      INTEGER idd, numb

#include "hbook.inc"

      INTEGER i4
      INTEGER jbit

      CALL hfind(idd, 'HNOENT')
      IF (IQUEST(1) .EQ. 0) THEN
         i4 = jbit(IQ(LCID + KBITS), 4)
         IF (i4 .NE. 0) THEN
            IQ(LCID + 3) = numb
         ELSE
            IQ(LCONT + KNOENT) = numb
         ENDIF
      ENDIF

      END SUBROUTINE
      
