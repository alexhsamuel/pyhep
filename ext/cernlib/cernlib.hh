//----------------------------------------------------------------------
//
// cernlib.hh
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

/* C++ Declarations of CERNLIB functions, and related utilities.  */

#ifndef __CERNLIB_HH__
#define __CERNLIB_HH__

//----------------------------------------------------------------------
// includes
//----------------------------------------------------------------------

#include <string>

//----------------------------------------------------------------------
// constants
//----------------------------------------------------------------------

#define PAWC_SIZE 4000000

//----------------------------------------------------------------------
// types
//----------------------------------------------------------------------

// Minuit callback function FCN(NPAR,GRAD,FVAL,XVAL,IFLAG,FUTIL).
typedef void 
(*mnfcn_t)(const int*, double*, double*, const double*, const int*, void*);


//----------------------------------------------------------------------
// function declarations
//----------------------------------------------------------------------

extern "C" {

// This is the common block which HBOOK uses as working space.
extern int pawc_[PAWC_SIZE];

// Function defined in cernlib.F.
extern void close_lun__(int* lun);
extern void hbook_init__();
extern void hfnoent_(int* idd, int* numb);
extern void open_lun__(const int* lun, const char* path, int* status, 
		       int len_path);

}  // extern "C"

//----------------------------------------------------------------------
// inline function definitions
//----------------------------------------------------------------------

/* Extract a string from a Fortran character buffer.

   'str' -- A whitespace-padded Fortran character buffer.

   'str_len' -- The length of the buffer.

   returns -- The string contents of the buffer, without any trailing
   whitespace. 
*/
inline std::string
stringFromFortran(const char* str,
		  size_t str_len)
{
  // Scan backwards, looking for non-spaces.
  for (const char* p = str + str_len - 1; p >= str; --p)
    // Found a non-space character.  The string extends up to here.
    if (*p != ' ')
      return std::string(str, p - str + 1);

  // The buffer is filled with spaces.  It's an empty string.
  return std::string("");
}


/* Fill a Fortran character buffer from a string.

   'source' -- The string with which to fill the buffer.

   'str' -- The character buffer.

   'str_length' -- The size of 'str'.

   If the length of 'source' exceeds 'str_length', it is truncated.
*/
inline void
makeFortranString(const std::string& source,
		  char* str,
		  size_t str_length)
{
  // Copy from the source into the buffer, truncating if necessary.
  size_t source_length = source.length();
  strncpy(str, source.c_str(), 
	  (source_length < str_length) ? source_length : str_length);
  // Right-pad the string with spaces, if it does not fill the buffer.
  if (source_length < str_length) 
    memset(str + source_length, ' ', str_length - source_length);
}


//----------------------------------------------------------------------
// cernlib Fortran functions
//----------------------------------------------------------------------

extern "C" {

extern void     dadmul_        (double (*f)(const int*, const double*),
				const int* n,
				const double* a,
				const double* b,
				const int* minpts,
				const int* maxptx,
				const double* eps,
				double* wk,
				const int* iwk,
				double* result,
				double* relerr,
				int* nfnevl,
				int* ifail);
extern double   dgauss_        (double (*f)(const double*), 
				const double* a, 
				const double* b, 
				const double* eps);
extern void     dinv_          (const int* n,
				double* a,
				const int* idim,
				int* ir,
				int* ifail);
extern void     hbarx_         (const int* id);
extern void     hbnamc_        (const int* id,
				const char* chblok,
				void* variable,
				const char* chform,
				int len_chblok,
				int len_chform);
extern void     hbname_        (const int* id,
				const char* chblok,
				void* variable,
				const char* chform,
				int len_chblok,
				int len_chform);
extern void     hbnt_          (const int* id,
				const char* chtitl,
				const char* chopt,
				int len_chtitl,
				int len_chopt);
extern void     hbook1_        (const int* id,
				const char* chtitl,
				const int* nx,
				const float* xmi,
				const float* xma,
				const float* vmx,
				int len_chtitl);
extern void     hbook2_        (const int* id,
				const char* chtitl,
				const int* nx,
				const float* xmi,
				const float* xma,
				const int* ny,
				const float* ymi,
				const float* yma,
				const float* vmx,
				int len_chtitl);
extern void     hbookn_        (const int* id,
				const char* chtitl,
				const int* nvar,
				const char* chrzpa,
				const int* nwbuff,
				const char* chtags,
				int len_chtitl,
				int len_chrzpa,
				int len_chtags);
extern int      hcbits_[37];
extern int      hcbook_[51];
extern void     hcdir_         (const char* chpath,
				const char* chopt, 
				int len_chpath,
				int len_chopt); 
extern void     hdcofl_        ();
extern void     hddir_         (const char* chpath,
				int len_chpath);
extern void     hdelet_        (const int* id);
extern void     hf1_           (const int* id,
				const float* x,
				const float* weight);
extern void     hfcxy_         (const int* icx,
				const int* icy,
				const float* x);
extern void     hfn_           (const int* id, 
				const char* x);
extern void     hfnt_          (const int* id);
extern void     hgive_         (const int* id,
				char* chtitl,
				int* nx,
				float* xmi,
				float* xma,
				int* ny,
				float* ymi,
				float* yma,
				int* nwt,
				void** loc,
				int len_chtitl);
extern void     hgiven_        (const int* id,
				char* chtitl,
				int* nvar,
				char* chtag,
				float* rlow,
				float* rhigh,
				int len_chtitl,
				int len_chtag);
extern void     hgnf_          (const int* id,
				const int* idnevt,
				char* x,
				int* ierror);
extern void     hgnpar_        (const int* id,
				const char* chrout,
				int len_chrout);
extern void     hgnt_          (const int* id,
				const int* idnevt,
				int* ierr);
extern void     hgntf_         (const int* id,
				const int* irow,
				int* ierr);
extern void     hgntv_         (const int* id,
				const char* chvar,
				const int* nvar,
				const int* irow,
				int* ierr,
				int len_chvar);
extern float    hi_            (const int* id,
				const int* i);
extern float    hij_           (const int* id,
				const int* i,
				const int* j);
extern float    hije_          (const int* id,
				const int* i,
				const int* j);
extern void     hkind_         (const int* id,
				int* kind,
				const char* chopt,
				int len_chopt);
extern void     hlnext_        (int* idh,
				char* chtype,
				char* chtitl,
				const char* chopt,
				int len_chtype,
				int len_chtitl,
				int len_chopt);
extern void     hmdir_         (const char* chpath,
				const char* chopt,
				int len_chpath,
				int len_chopt);
extern void     hnoent_        (const int* id,
				int* noent);
extern bool     hntnew_        (const int* id);
extern void     hntvdef_       (const int* id1,
				const int* ivar,
				char* chtag,
				char* block,
				int* itype,
				int len_chtag,
				int len_block);
extern void     hpak_          (const int* id,
				float* conten);
extern void     hpake_         (const int* id,
				float* conten);
extern void     hrdir_         (const int* maxdir,
				char** chdir,
				int* ndir,
				int len_chdir);
extern void     hrend_         (const char* chopt,
				int len_chopt);
extern void     hrin_          (int* id,
				int* icycle,
				int* iofset);
extern void     hropen_        (const int* lun, 
				const char* chtop,
				const char* chfile, 
				const char* chopt, 
				int* lrec, 
				int* istat, 
				int len_chtop, 
				int len_chfile,
				int len_chopt);
extern void     hrout_         (const int* id, 
				int* icycle,
				const char* chopt,
				int len_chopt);
extern void     hscr_          (const int* id,
				const int* icycle,
				const char* chopt,
				int len_chopt);
extern void     hunpak_        (const int* id,
				float* conten,
				const char* choice,
				int* num,
				int len_choice);
extern void     hunpke_        (const int* id,
				float* conten,
				const char* choice,
				int* num,
				int len_choice);
extern void     mncomd_        (mnfcn_t fcn, 
				const char* chstr, 
				int* icondn,
				void* futil, 
				int len_chstr);
extern void     mnemat_        (double* emat,
				const int* ndim);
extern void     mnerrs_        (const int* num,
				double* eplus,
				double* eminus,
				double* eparab,
				double* globcc);
extern void     mninit_        (const int* ird, 
				const int* iwr, 
				const int* isav);
extern void     mnparm_        (const int* num, 
				const char* chnam, 
				const double* stval, 
				const double* step, 
				const double* bnd1, 
				const double* bnd2, 
				int* ierflg, 
				int len_chnam);
extern void     mnpout_        (const int* num,
				char* chnam,
				double* val,
				double* error,
				double* bnd1,
				double* bnd2,
				int* ivarbl,
				int len_chnam);
extern void     mnstat_        (double* fmin,
				double* fedm,
				double* errdef,
				int* npari,
				int* nparx,
				int* istat);
extern float    prob_          (float* x,
				int* n);
extern int      quest_[100];
extern int      rzcl_[11];
extern void     rzlogl_        (const int* lun,
				const int* loglev);
extern void     rzpurg_        (const int* nkeep);
extern void     rzvout_        (const int* vect, 
				const int* nout, 
				const int* key,
				int* icycle, 
				const char* chopt, 
				int len_chopt);

}  // extern "C"

//----------------------------------------------------------------------

#endif  // #ifndef __CERNLIB_HH__
