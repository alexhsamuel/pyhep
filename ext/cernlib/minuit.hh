//----------------------------------------------------------------------
// 
// minuit.hh
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

/* Python interface to the MINUIT library.  */

#ifndef __MINUIT_HH__
#define __MINUIT_HH__

//----------------------------------------------------------------------
// includes
//----------------------------------------------------------------------

#include "python.hh"

//----------------------------------------------------------------------
// function declarations
//----------------------------------------------------------------------

extern PyObject* function_minuit_minimize(PyObject* self, Py::Arg* args);
extern PyObject* function_minuit_maximumLikelihoodFit(PyObject* self, 
						      Py::Arg* args);

//----------------------------------------------------------------------

#endif  // #ifndef __MINUIT_HH__
