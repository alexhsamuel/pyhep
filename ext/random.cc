//----------------------------------------------------------------------
//
// random.cc
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

/* Random number generation functions.  */

//----------------------------------------------------------------------
// includes
//----------------------------------------------------------------------

#include <cassert>
#include <cstddef>
#include <sys/time.h>

#include "random.hh"

//----------------------------------------------------------------------
// constants
//----------------------------------------------------------------------

namespace {

// Constants for the first congruential generator.
const static long m1 = 2147483563;
const static long a1 = 40014;
const static long q1 = 53668;
const static long r1 = 12211;

// Constants for the second congruential generator.
const static long m2 = 2147483399;
const static long a2 = 40692;
const static long q2 = 52774;
const static long r2 = 3791;

//----------------------------------------------------------------------
// helper functions
//----------------------------------------------------------------------

inline float
min(float v1,
    float v2)
{
  return (v1 < v2) ? v1 : v2;
}


}  // anonymous namespace

//----------------------------------------------------------------------
// methods
//----------------------------------------------------------------------

ShuffledLEcuyerRandom::ShuffledLEcuyerRandom(long seed)
{
  initialize(seed);
}


void
ShuffledLEcuyerRandom::initialize(long seed)
{
  // If no seed was specified, construct one from the system time.
  if (seed == 0) {
    struct timeval now;
    int result = gettimeofday(&now, NULL);
    assert(result == 0);
    seed = now.tv_sec ^ now.tv_usec;
  }

  // Initialize.
  assert(seed > 0);
  state1_ = seed;
  state2_ = seed;

  // Load the shuffle table (after eight warm-ups).
  for (long j = table_size_ + 7; j >= 0; j--) {
    long k = state1_ / q1;
    state1_ = a1 * (state1_ - k * q1) - k * r1;
    if (state1_ < 0)
      state1_ += m1;
    if (j < table_size_)
      shuffle_table_[j] = state1_;
  }
  last_ = shuffle_table_[0];
}


float
ShuffledLEcuyerRandom::random()
{
  long k;
  static float inv_m1 = 1.0 / m1;

  // Next value from the first congruential generator.
  k = state1_ / q1;
  state1_ = a1 * (state1_ - k * q1) - k * r1;
  if (state1_ < 0)
    state1_ += m1;

  // Next value from the second congruential generator.
  k = state2_ / q2;
  state2_ = a2 * (state2_ - k * q2) - k * r2;
  if (state2_ < 0)
    state2_ += m2;

  // Combine and shuffle.
  int j = last_ / (1 + (m1 - 1) / table_size_);
  last_ = shuffle_table_[j] - state2_;
  shuffle_table_[j] = state1_;
  if (last_ < 1)
    last_ += m1 - 1;

  // Make sure we never return 1.0.
  return min(last_ * inv_m1, 1.0 - 1.2e-7);
}


