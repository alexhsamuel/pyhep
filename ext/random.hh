//----------------------------------------------------------------------
//
// random.hh
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

/* Random number generation functions.  */

#ifndef __RANDOM_HH__
#define __RANDOM_HH__

//----------------------------------------------------------------------
// classes
//------------------------------------------------------------------------

/* Shuffled L'Ecuyer random number generator.

   Generator for uniform deviates based on 

     L'Ecuyer, P. 1988, Communications of the ACM, vol. 31, pp. 742-774

   as described and implemented as the 'ran2' function in

     Press, William H. et al., Numerical Recipes in C, Cambridge
     University Press, 1988, p. 282.

   This generator is based on two multiplicative congruential generators
   with factors 'a1' and 'a2' and moduli 'm1' and 'm2, with
   multiplication performed with Schrage's algorith.  The values from
   the two sequences are combined, and the sequence is shuffled using a
   table of 'tab_size'.  

   The period of the generator is about 2.3 x 10^18.

*/
class ShuffledLEcuyerRandom
{
public:

  /* Create a new generator.

     'seed' -- A positive integer seed value.  If zero, a seed is chosen
     using the system time. 
  */
  ShuffledLEcuyerRandom(long seed=0);
  
  /* Initialize the generator. 

     'seed' -- A positive integer seed value.  If zero, a seed is chosen
     using the system time. 
  */
  void initialize(long seed=0);

  /* Return a uniform deviate between 0 inclusive and 1 exclusive.  */
  float random();

private:

  // Forbid copies.
  ShuffledLEcuyerRandom(const ShuffledLEcuyerRandom&);

  // Shuffle table size.
  const static int table_size_ = 32;

  // State variables for the congruential generators.
  long state1_;
  long state2_;

  // State of the combined generator; used for shuffling.
  long last_;

  // The shuffle table.
  long shuffle_table_[table_size_];

};

//----------------------------------------------------------------------

#endif  // #ifndef __RANDOM_HH__
