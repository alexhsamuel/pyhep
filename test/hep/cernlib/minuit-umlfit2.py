#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep.cernlib.minuit import maximumLikelihoodFit
import hep.table
from   hep.test import compare
import random

#-----------------------------------------------------------------------
# test
#-----------------------------------------------------------------------

schema = hep.table.Schema()
schema.addColumn("x", "float32")
table = hep.table.create("minuit-umlfit2.table", schema)
for i in xrange(100):
    table.append(x=random.normalvariate(3, 1))

result = maximumLikelihoodFit(
    "gaussian(mu, sigma, x)", (("mu", 0), ("sigma", 10, 0.1, 0, 1000)),
    table)
print result

compare(result.minuit_status, 3)
compare(result.values["mu"], 3, precision=0.3)
compare(result.values["sigma"], 1, precision=0.3)
