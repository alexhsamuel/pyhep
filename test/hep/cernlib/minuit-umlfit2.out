Traceback (most recent call last):
  File "minuit-umlfit2.py", line 22, in ?
    table)
  File "/home/samuel/dev/pyhep/py/hep/cernlib/minuit.py", line 239, in maximumLikelihoodFit
    parameters = normalizeParameters(parameters)
  File "/home/samuel/dev/pyhep/py/hep/minimize.py", line 42, in normalizeParameters
    bound_lo, bound_hi = map(float, parameter[3])
TypeError: argument 2 to map() must support iteration
